test/tools/router_check/router.cc

```diff
@@ -126,17 +127,44 @@ ToolConfig::ToolConfig(std::unique_ptr<Http::TestRequestHeaderMapImpl> request_h
       random_value_(random_value) {}
 
 // static
-RouterCheckTool RouterCheckTool::create(const std::string& router_config_file,
-                                        const bool disable_deprecation_check) {
-  // TODO(hennna): Allow users to load a full config and extract the route configuration from it.
+RouterCheckTool RouterCheckTool::create(const std::string& config_file,
+                                        const bool disable_deprecation_check,
+                                        const std::string& listener_name) {
   envoy::config::route::v3::RouteConfiguration route_config;
+  envoy::config::bootstrap::v3::Bootstrap bootstrap_config;
+
   auto stats = std::make_unique<Stats::IsolatedStoreImpl>();
   auto api = Api::createApiForTest(*stats);
-  TestUtility::loadFromFile(router_config_file, route_config, *api);
-  assignUniqueRouteNames(route_config);
-  assignRuntimeFraction(route_config);
+  try {
+    TestUtility::loadFromFile(config_file, route_config, *api);
+  } catch (const EnvoyException& _) {
+    TestUtility::loadFromFile(config_file, bootstrap_config, *api);
+  }
+
   auto factory_context =
       std::make_unique<NiceMock<Server::Configuration::MockServerFactoryContext>>();
+
+  auto extracted_route_config =
+      RouterCheckTool::extractRouteConfigFromListener(bootstrap_config, listener_name);
+  if (extracted_route_config.has_value()) {
+    route_config = extracted_route_config.value();
+  }
+
+  if (bootstrap_config.has_layered_runtime()) {
+    absl::StatusOr<std::unique_ptr<Runtime::LoaderImpl>> loader =
+        Envoy::Runtime::LoaderImpl::create(
+            factory_context->dispatcher_, factory_context->thread_local_,
+            bootstrap_config.layered_runtime(), factory_context->local_info_,
+            factory_context->api_.stats_store_, factory_context->api_.random_,
+            factory_context->validation_context_.dynamicValidationVisitor(),
+            factory_context->api());
+    if (!loader.ok()) {
+      std::cerr << "Failed to initialize runtime: " << loader.status().ToString() << std::endl;
+    }
+  }
+
+  assignUniqueRouteNames(route_config);
```

Comment:
It seems there could be paths that leas to an invalid/empty route_config (e.g., if `extractRouteConfigFromListener()` returns nullopt.

As a high-level comment, IMHO it would be better to have the code-paths for bootstrap and route processing separate. Specifically, one can look whether the given config is of type Bootstrap or Route (probably by first converting to Any), and handle that accordingly.

Reply:
Thank you for the feedback @adisuissa . I agree with having different code paths according to the given config and I'm trying to get this working. I failed to find a reference of reading a config as Any and checking if Is one or the other type (I'm still new to the language and codebase). Do you have a pointer for this? I tried something on those lines but it clearly did not work.


```cpp
  auto stats = std::make_unique<Stats::IsolatedStoreImpl>();
  auto api = Api::createApiForTest(*stats);

  ProtobufWkt::Any any_config;
  TestUtility::loadFromFile(config_file, any_config, *api);
```

A simpler approach would be to use a separate flag `--bootstrap-config`.